### PR TITLE
Add CycloneDDS packages to Dashing

### PR DESCRIPTION
**Pull request by cottsay**
These packages haven't been released yet, but should be soon.

### CODE REVIEW

**Comment by nuclearsandwich** [comment]
The dependencies listed for rmw_cyclonedds have dependencies on `CycloneDDS` which is not going to resolve properly. I think those keys should be changed to `cyclonedds` to match the expected ROS package name from the release repository.

https://github.com/atolab/rmw_cyclonedds/blob/master/rmw_cyclonedds_cpp/package.xml#L15-L29

**Comment by cottsay** [comment]
`CycloneDDS` is the name of the CMake project. colcon detects it properly like this. If we change the reference in `rmw_cyclonedds`, we'll need to override the package name in the CycloneDDS package with either a `package.xml` or a `colcon.pkg` (I'd prefer the latter). Then we can patch `rmw_cyclonedds`.

Thoughts from @dirk-thomas?

**Comment by nuclearsandwich** [comment]
There are a few cases of capitals in distribution.yaml entries. It seems then that the keys just need to agree so that rosdep and colcon can both find the dependency.

```
❯ rg '^  \w*[A-Z]\w*:'  **/distribution.yaml
hydro/distribution.yaml
6351:  rFSM:
6923:  rosR:
6928:  rosR_demos:

groovy/distribution.yaml
16:  RCPRG_laser_drivers:
3854:  rosR:
3859:  rosR_demos:

kinetic/distribution.yaml
6415:  lslidar_C16:

indigo/distribution.yaml
12573:  rosR:
12583:  rosR_demos:
```

**Comment by dirk-thomas** [comment]
> There are a few cases of capitals in distribution.yaml entries.

The problem is not the entry in the `yaml`. The package name must be lowercase.

**Comment by dirk-thomas** [comment]
Since we don't have write access to the upstream repo you can't place any files there. In the release repo a `colcon.pkg` file doesn't make any sense since it is not processed by `colcon` but `dpkg-buildpackage`.

Therefore I think you need to use the lower case name `cyclonedds` in the `package.xml` file which is added in the release repo.

As a consequence the rmw package must use that lowercase name as its dependency (which will make it not work in a from-source build).

**Comment by cottsay** [comment]
>  ...we should set a known good commit as the source version.

I'll update the PR.

> ...you need to use the lower case name `cyclonedds` in the `package.xml`...

Makes sense, I'll do it that way.

> ...request to add a `colcon.pkg` file into the `CycloneDDS` repo...

I agree, that sounds like the best thing to do.

File: dashing/distribution.yaml
Comment (cottsay) [reply dashing/distribution.yaml]: What's the consensus here? I added them only because I saw them on other entries.

File: dashing/distribution.yaml
Comment (cottsay) [reply dashing/distribution.yaml]: Done - snapped to the HEAD of each, which I confirmed builds successfully locally.

File: dashing/distribution.yaml
Comment (dirk-thomas) [reply dashing/distribution.yaml]: Please reference those here since they should be removed too.

Comment (cottsay) [reply dashing/distribution.yaml]: > Please reference those here since they should be removed too.

I'm not sure what you mean.

Here is the check that fails unless these are set when specifying a commit: https://github.com/ros/rosdistro/blob/158a6037e658f8196e979969e9c41031a355a493/test/test_url_validity.py#L131-L133

Comment (dirk-thomas) [reply dashing/distribution.yaml]: Oh, I didn't know about that unconditional check in Travis check in this repo. Then I guess it must stay.

**Comment by dirk-thomas** [comment]
Are you going to change the dependency in the rmw package in the release repo then?

**Comment by cottsay** [comment]
> Are you going to change the dependency in the rmw package in the release repo then?

Yes, kind-of. At least right now, the last commit in the release repo is the one that changes it to uppercase. So I'll just target that commit instead of patching. For now.

**Comment by cottsay** [comment]
I added release info for the cyclonedds GBP to this PR. Merging it should unblock debs for that package.

I'm working on the RMW packages now.

**Comment by cottsay** [comment]
Note that we should merge #22116 before this PR.

**Review by dirk-thomas — APPROVED** [review]

**Pull request merged**
